### DISCUSSION TITLE
How to know when rendering is complete?

### DISCUSSION THREAD

**Discussion by sam-wheat** [microsoft/microsoft-ui-xaml#8362] [Q&A]
My app gets data from a database, sorts and groups data, and builds a pretty complex control in code.  This is done in about a half second - however rendering the control takes WinUI eight seconds or more.  I can display a progress ring while my control is working however the user does not see it for the half second it is visible and they wind up seeing a frozen UI for the 7.5 seconds while WinUI is rendering. How do I display the progress ring and hide it only when the page is rendered and ready to use?

**Comment by DarranRowe** [comment]
The big question I have is how much of this work are you pushing off onto a worker thread?
Remember, if you are doing this rendering on the main UI thread, you are going to clog up the event handling of the UI and this means that your application won't have the chance to do anything but show a frozen UI.

**Reply by sam-wheat** [reply]
I am not controlling rendering at all.  In my code I use `GetTemplateChild` to find a grid then I add a bunch of rows/columns.  That part takes a half second.  Everything after that - WinUI rendering the screen - is what is taking so long.

My control is [here](https://github.com/leaderanalytics/LeaderPivot.XAML.WinUI).

Edit:

For reference, the [same control running on Blazor](https://leaderanalytics.com/blazor/leader-pivot-demo) is nearly instantaneous.  Same control that loads data and constructs the matrix, same test data.

**Reply by nikolayvpavlov** [reply]
Do you use virtualization?

**Reply by sam-wheat** [reply]
No

**Reply by nikolayvpavlov** [reply]
This can be a serious issue.  Xaml is notorious for being slow computing large visual trees. :-(

**Reply by sam-wheat** [reply]
I'd love to speed it up - any tips you can send my way will be much appreciated.  In the meantime, do you know any way I might handle some kind of "render complete" event so I can put up a progress ring while the UI is rendering?

**Reply by castorix** [reply]
> I'd love to speed it up - any tips you can send my way will be much appreciated. In the meantime, do you know any way I might handle some kind of "render complete" event so I can put up a progress ring while the UI is rendering?

I use [CompositionTarget.Rendering](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.media.compositiontarget.rendering?view=windows-app-sdk-1.2) with WinUI3/Direct2D (to render at 60 FPS with my monitor), but not sure if it can be useful in your case...

**Comment by sam-wheat** [comment]
Microsoft?  Any response to this?

**Comment by roxk** [comment]
From your description, it seems you are adding a bunch of views to the visual tree. You should use list view to virtualize your rows/columns.
